core/structs-checker/src/lib.rs

```diff
@@ -0,0 +1,103 @@
+use proc_macro::TokenStream;
+use quote::ToTokens;
+use quote::{format_ident, quote};
+use syn::{parse_macro_input, Data, DeriveInput, Fields, FieldsNamed, FieldsUnnamed, Variant};
+
+#[proc_macro_derive(ProtocolStruct)]
+pub fn protocol_struct(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    let name = &input.ident;
+    let info_name = format_ident!("{}_INFO", name);
+
+    let info = match &input.data {
+        Data::Struct(data_struct) => {
+            let fields = extract_struct_fields(&data_struct.fields);
+            quote! {
+                near_structs_checker_core::ProtocolStructInfo::Struct {
+                    name: stringify!(#name),
+                    fields: &[#(#fields),*],
+                }
+            }
+        }
+        Data::Enum(data_enum) => {
+            let variants = extract_enum_variants(&data_enum.variants);
+            quote! {
+                near_structs_checker_core::ProtocolStructInfo::Enum {
+                    name: stringify!(#name),
+                    variants: &[#(#variants),*],
+                }
+            }
+        }
+        Data::Union(_) => panic!("Unions are not supported"),
+    };
+
+    let expanded = quote! {
+        #[allow(non_upper_case_globals)]
+        static #info_name: near_structs_checker_core::ProtocolStructInfo = #info;
+
+        inventory::submit!(#info_name);
+
+        impl near_structs_checker_core::ProtocolStruct for #name {}
+    };
+
+    TokenStream::from(expanded)
+}
+
+fn extract_struct_fields(fields: &Fields) -> Vec<proc_macro2::TokenStream> {
+    match fields {
+        Fields::Named(FieldsNamed { named, .. }) => named
+            .iter()
+            .map(|f| {
+                let name = &f.ident;
+                let ty = &f.ty;
```

Comment:
One thing to worry about: 

```rust
type Foo = u64; // can be changed to something else without affecting outcome of `stringify!()`.

#[derive(ProtocolStruct)]
struct ProtocolSchema {
    a: Foo,
}
```

In this aspect a type id is quite a bit more sound. I'd say that instead of trying to come up with our own way to create type identifiers, lets keep using `TypeId::of`, which apparently can be used in const contexts with a nightly compiler (or `env RUSTC_BOOTSTRAP=1`): https://github.com/rust-lang/rust/issues/77125.

We already depend on a nightly compiler for some checks, so it wouldn't be too terrible of an idea to do it here as well, especially since we do plan to put this behind a feature gate among other things.

(You may still need to keep track of field names to detect when a field swap occurs.)